File: cpp/ros2/src/ngg/src/subscriber/subscriber_member_function.cpp
// Copyright 2019 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "subscriber_member_function.hpp"

MinimalSubscriber::MinimalSubscriber()
: Node("minimal_subscriber"), game_setting(43, 7), game_state()
{
  subscription_ = this->create_subscription<std_msgs::msg::String>(
    "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
  updater_ = std::make_unique<diagnostic_updater::Updater>(this);
  updater_->setHardwareID("number_guessing_game");
  updater_->add("game_status", this, &MinimalSubscriber::produce_diagnostics);
}

void MinimalSubscriber::produce_diagnostics(diagnostic_updater::DiagnosticStatusWrapper & stat)
{
  if (game_state.is_success_) {
    stat.summary(diagnostic_msgs::msg::DiagnosticStatus::OK, "Game completed successfully");
  } else if (game_state.current_attempt_ >= game_setting.MAX_ATTEMPTS) {
    stat.summary(diagnostic_msgs::msg::DiagnosticStatus::ERROR, "Game over - max attempts reached");
  } else {
    stat.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "Game in progress");
  }
}

void MinimalSubscriber::topic_callback(const std_msgs::msg::String & msg)
{
  RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.data.c_str());
  if (game_state.current_attempt_ < game_setting.MAX_ATTEMPTS && !game_state.is_success_) {
    try {
      std::string input = msg.data;
      int num = std::stoi(input);
      game_state = game_state.update_state(num, game_state, game_setting);
      if (game_state.is_success_) {
        std::cout << "Correct" << std::endl;
      } else {
        std::cout << "Incorrect" << std::endl;
      }
      updater_->force_update();
    } catch (const std::invalid_argument &) {
      // 例外オブジェクトを使わない場合は、const参照を使わなくてもいい。
      std::cout << "Invalid Input" << std::endl;
    } catch (const std::out_of_range & e) {
      std::cout << "Out of range" << std::endl;
    }
  }
  if (game_state.current_attempt_ == game_setting.MAX_ATTEMPTS) {
    std::cout << "max attempts" << std::endl;
    updater_->force_update();
  }
}

GameState GameState::update_state(int input, GameState game_state, GameSetting game_setting)
{
  return GameState(
    input, game_state.current_attempt_ + 1, game_setting.SECRET_NUM == input ? true : false);
}

File: python/ros2/src/ngg/ngg/subscriber/subscriber_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

class MinimalSubscriber(Node):
    def __init__(self, game_manager):
        super().__init__('minimal_subscriber')
        self._game_manager = game_manager
        self.subscription_ = self.create_subscription(
                String,
                'topic',
                self.listener_callback,
                10)
        self.subscription_

    def listener_callback(self, msg):
        self.get_logger().info(f'I heard: {msg.data}')
        self._game_manager.process_data(msg.data)

class GameSetting:
    def __init__(self, secret_number, max_attempts):
        self._secret_number = secret_number
        self._max_attempts = max_attempts

class GameState:
    def __init__(self, input=0, current_attempt=0, is_success=False):
        self._last_input_number = input
        self._current_attempt = current_attempt
        self._is_success = is_success

    def update_state(self, input, game_setting):
        print(input)
        print(game_setting._secret_number==input)
        return GameState(
                input,
                self._current_attempt+1,
                game_setting._secret_number==input)

class GameManager:
    def __init__(self):
        self._game_setting = GameSetting(43, 7)
        self._game_state = GameState()

    def str_to_int(self, input):
        try:
            return int(input)
        except ValueError:
            print(f'{input}はintに変換できません')
            return None
    def process_data(self, input):
        input_int = self.str_to_int(input)
        if input_int:
            self._game_state = self._game_state.update_state(input_int, self._game_setting)
            if self._game_state._is_success:
                print("正解です")
            else:
                print("不正解です")




def main(args=None):
    rclpy.init(args=args)
    game_manager = GameManager()
    minimal_subscriber = MinimalSubscriber(game_manager)
    rclpy.spin(minimal_subscriber)
    minimal_subscriber.destroy_node()
    rclpy.shutdown()
